File: frontend/src/lib/online/game-response.ts
export type Player = {
  id: number,
  sub: string,
  email: string,
  name: string,
  username: string,
  picture: string
};

export type GameResponse = {
  id: string,
  uciFen: string,
  whitePlayer: Player,
  blackPlayer: Player,
  whiteElo: 0.1,
  blackElo: 0.1,
  whiteEloChange: 0.1,
  blackEloChange: 0.1,
  startTime: string,
  endTime: string,
  blackTimeLeft: number,
  whiteTimeLeft: number,
  whiteOfferingDraw: boolean,
  blackOfferingDraw: boolean,
  result: string,
  resultDetail: string
  createdAt: string,
  isGameWithBot: boolean,
  isStarted: boolean,
  gameTypeId: number,
};

File: backend/src/main/java/com/se330/ctuong_backend/dto/rest/GameResponse.java
package com.se330.ctuong_backend.dto.rest;

import com.se330.ctuong_backend.config.ApplicationConfiguration;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.sql.Timestamp;
import java.time.Instant;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class GameResponse {
    private String id;
    private String uciFen;
    private User whitePlayer;
    private User blackPlayer;

    private Long blackElo;
    private Long whiteElo;

    private Long whiteEloChange;
    private Long blackEloChange;

    private Timestamp startTime;
    private Instant endTime;
    private Boolean isEnded;
    private Timestamp createdAt;

    private Long gameTypeId;

    @NotNull
    private Long blackTimeLeft;

    @NotNull
    private Long whiteTimeLeft;

    private String result;
    private String resultDetail;
    private Boolean isStarted;

    public boolean getIsGameWithBot() {
        return whitePlayer != null && whitePlayer.getId().equals(ApplicationConfiguration.getBotId()) ||
                blackPlayer != null && blackPlayer.getId().equals(ApplicationConfiguration.getBotId());
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class User {
        private Long id;
        private String sub;
        private String email;
        private String name;
        private String username;
        private String picture;
    }
}
